Enhancement of descriptor class

ManuelSchauer opened the pull request:
Enhancement of parallelization strategies and other small changes.

ManuelSchauer committed:
changed fingerprint calculation and added single molecule smiles calculation, also moved checks into validation methods

ManuelSchauer committed:
added exception handling to fingerprint pool initialization and added single molecule smiles calculation to parallelization test

JonasSchaub commented:
@ManuelSchauer please add your name in first place to the License-header/License-header.txt file and then execute the Gradle command spotlessApply to apply this change to all files.

JonasSchaub commented:
@ManuelSchauer please add Java versions 8 and 25 to the .github/workflows/gradle.yml file (line 25), so that we can already test for compatibility with these two versions.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -98,21 +98,26 @@
  * Descriptor related calculations based on the CDK for the enrichment of data vectors.
- * Note: There are 3 "public static boolean setDescriptorsForMolecules...()" methods with different forms of
- * (parallelized) calculation since individual CDK descriptor calculation classes are unfortunately NOT thread-safe.
+ * Note: There are 4 "public static boolean setDescriptorsForMolecules...()" methods with different forms of
```
I'm counting 5 methods. Simply omit the exact number, I would say.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -98,21 +98,26 @@
+ * (parallelized) calculation. There is single molecule processing or batch processing for large datastructures.
+ * Single and Batch processing can be used with either an IAtomContainer array or with a String Array of SMILES codes.
  * Note: For adding a new descriptor, go to "Add new descriptor information here!"
  *
```
add `<p>` in the empty line, so that it is represented correctly in the formatted JavaDoc.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -121,7 +126,12 @@
- * // 4. Define which descriptors to calculate by creating a Descriptor array
+ * // B: String preprocessing
+ * // 1: Create a String array of SMILES codes
+ * String[] moleculeSmilesStrings = "CCO", "CCC(O)O", ...
```
on the right-hand side of equals, it needs to be new String[] {"CCO", "CCC(O)O", ...};

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -158,11 +179,18 @@
+ *  *
  * }</pre>
  *
  * @author Achim Zielesny
```
also add tag `@version 1.0.0` here.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -1106,17 +1134,7 @@ void calculate(IAtomContainer anAtomContainer, float[] aVector, int aStartIndex)
     PUBCHEM_FINGERPRINTER(true, true, true, false, 881) {
         @Override
         void calculate(IAtomContainer anAtomContainer, float[] aVector, int aStartIndex) {
```
(I'm marking this here because I cannot comment above) [nitpick] Here, it is not really "aStartIndex" like in the method above, where it really defines from which matrix column on, the descriptors should be entered, but rather the specific index where this specific descriptor should be entered into the matrix. I would suggest changing the param name here simply to "index".

JonasSchaub commented:
@ManuelSchauer an alternative way to having the calculate() method defined in every enum constant, one could think of implementing one central method that gets the constant as parameter, queries its properties (e.g. whether it needs explicit hydrogens), does preprocessing accordingly if necessary, gets the associated object from the map, does the calculation, checks the type of the produced descriptor result (e.g. instanceof IntegerResult, IntegerArrayResult, etc.), and enters the generated values into the map accordingly. This is the reason why there is this complex API with the different descriptor results classes. From our point of view, what would be the disadvantages of that? I mean, I like how ordered and clearly defined our way is, but probably, we will have to defend it when we try to get it into CDK. Maybe we should discuss this in the new year.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -1106,17 +1134,7 @@ void calculate(IAtomContainer anAtomContainer, float[] aVector, int aStartIndex)
     PUBCHEM_FINGERPRINTER(true, true, true, false, 881) {
```
isFast = true, really? 881 substructure searches are fast enough to be qualified as "fast" here? Did you test that?

JonasSchaub replied:
I feel like we discussed this before. If true, I apologise for asking this again.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -1439,9 +1367,16 @@ public int getDescriptorComponentNumber() {
      */
```
(cannot comment above) Descriptor constructor doc: param tags for isFingerprint and needsExplicitHydrogens are missing.

JonasSchaub replied:
Also: methods like isFingerprint() or isFast() could have some additional documentation. E.g. isFingerprint() should document that enum constants where this field is true cannot be used for certain types of parallel computation. And what does "fast" mean exactly?

JonasSchaub replied:
Also, there are 2 unnecessary empty lines after getDescriptorComponentNumber()

JonasSchaub commented:
Also, what were our reasons to keep the descriptorToCdkObjectMap? So that we can do exception handling for the descriptor class initialisations? Was there an additional reason?

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -1822,6 +1832,40 @@ public static int getNumberOfComponents(
         }
```
comment on getSpecifiedDescriptors(): I suggest removing the isQuicklyCalculableDescriptorInclusion parameter to make the overall concept cleaner. The other parameters are ex-/including descriptors with unfavourable properties (unsafe, slow, fingerprints). I don't think there is a use case where you only want the slow descriptors.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -1665,42 +1600,117 @@ public int getDescriptorComponentNumber() {
+     * This method creates new BlockingQueues for each fingerprint type and populates them
+     * with fingerprinter instances. Thread-safe for concurrent access.
+     */
+    private static synchronized void initializeFingerprintPools() {
```
only vaguely related to this: can the SMILESParser be a bottle neck? Would it make sense to have a pool of SmilesParser instances as well?

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -3317,64 +2655,6 @@ private static boolean setDescriptor(
-
-                // Check for NaN values in the calculated result and track them
-                int numComponents = aDescriptor.getDescriptorComponentNumber();
-                return !checkAndTrackNaNValues(aVector, aStartIndex, numComponents, aMoleculeIndex, aNanPositions);
```
why is checkAndTrackNaNValues() package-private?

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -3925,21 +3191,155 @@ static boolean checkAndTrackNaNValues(float[] aVector, int aStartIndex, int numC
+    }
+
+    /**
+     * Validates an atom container input and its molecules.
```
is there an "array" missing in the sentence?

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -3925,21 +3191,155 @@ static boolean checkAndTrackNaNValues(float[] aVector, int aStartIndex, int numC
+    }
+
+    /**
+     * Validates an atom container input and its molecules.
```
adjust validateSmilesStringArray() doc.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -3925,21 +3191,155 @@ static boolean checkAndTrackNaNValues(float[] aVector, int aStartIndex, int numC
-        float[] floatArray = new float[size];
+    private static void validateBatchSize(String methodName, int aBatchSize) {
+        if (aBatchSize <= 0) {
+        throw new IllegalArgumentException(methodName + ": aBatchSize must be greater than 0.");
```
correct indentation.

JonasSchaub reviewed (commented):
@ManuelSchauer I could not do an in-depth review before the Christmas break, sorry. Here are my few cents to your code. Overall, everything is looking great. I'm looking forward to finalising this project in the new year.

ManuelSchauer commented:
> @ManuelSchauer an alternative way to having the calculate() method defined in every enum constant, one could think of implementing one central method that gets the constant as parameter, queries its properties (e.g. whether it needs explicit hydrogens), does preprocessing accordingly if necessary, gets the associated object from the map, does the calculation, checks the type of the produced descriptor result (e.g. instanceof IntegerResult, IntegerArrayResult, etc.), and enters the generated values into the map accordingly. This is the reason why there is this complex API with the different descriptor results classes. From our point of view, what would be the disadvantages of that? I mean, I like how ordered and clearly defined our way is, but probably, we will have to defend it when we try to get it into CDK. Maybe we should discuss this in the new year.

I do not see any disadvantages right now. I will look into it so that we can discuss this next year.

ManuelSchauer committed:
Updated documentation, and License-header.txt, small changes to error handling

JonasSchaub commented:
@ManuelSchauer please ignore what I tried in https://github.com/JonasSchaub/CDK-Descriptor-Calculation/pull/8, it seems to be impossible to make both Java 8 and 25 work here.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -158,16 +180,24 @@
- * // Parameters are identical to Option A
+ * // Option B: High-Performance-Parallel Descriptor Calculation with SMILES string array
+ *  * boolean result = Descriptor.setDescriptorsForMoleculesBySmilesStringParallelization(
+ *  *                 descriptors,
```
the second column of asterisks doesn't render properly/nicely. I would simply remove it.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -1665,42 +1602,117 @@ public int getDescriptorComponentNumber() {
+        BlockingQueue<IFingerprinter> pubchemPool = new LinkedBlockingQueue<>(fingerprintPoolSize);
+        for (int i = 0; i < fingerprintPoolSize; i++) {
+            if (!pubchemPool.offer(new PubchemFingerprinter(SilentChemObjectBuilder.getInstance()))) {
+                throw new RuntimeException("Failed to initialize pubchem fingerprint pool.");
```
If I understand this correctly, offer() only returns false if the queue is already full, which should never happen here.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -2259,12 +2119,7 @@ public static boolean setDescriptorsForMoleculeSmilesStringsByBatchParallelizati
-                    anException
-            );
-            throw new Exception("Descriptor.setDescriptorsForMoleculeStringsByBatchParallelization: An exception occurred.", anException);
+            throw new Exception("Descriptor.setDescriptorsForMoleculeBySmilesStringsBatchParallelization: An exception occurred.", anException);
```
rather throw CDKException here instead of the generic exception.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -1822,6 +1834,40 @@ public static int getNumberOfComponents(
+
     /**
      * Uses a specified aromaticity model to modify aMolecule.
      * Note: This method changes the input molecule by applying the specified aromaticity model.
```
This method does more than the doc tells you, please extend the doc.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -2453,197 +2221,110 @@ public static boolean setDescriptorsForMoleculesByMoleculeParallelization(
-                    Level.SEVERE,
-                    "Descriptor.setDescriptorsForMoleculesByMoleculeParallelization: An exception occurred: This should never happen.", anException
-            );
             throw anException;
```
catching and re-throwing the same exception should not be done.

ManuelSchauer committed:
Updated calculation method and minor doc changes.

JonasSchaub commented:
@ManuelSchauer, the issue with spotless (or rather: the build on Java 11) is more complex (as always), apparently. 

Why it arose is because you are now using pattern matching for instanceof (e.g. "if (result instanceof DoubleResult doubleResult) {..."), which was only introduced in Java 14! 

This, of course, made me curious as to why the code was compiling, then. Apparently, spotless uses the actual Java version given in the workflow file, while Gradle always uses Java 21 because we (I) defined this in the build file as Java toolchain version:
```
java {
    toolchain {
        languageVersion = JavaLanguageVersion.of(21)
    }
    // ...
}
```

So, I would ask you to remove this from the build file, not use pattern matching for instanceof, and then see whether all the build are working (might be that we spot other incompatibilities with the individual Java versions then).

ManuelSchauer committed:
Implemented getDescriptorAndComponentIndex and getDescriptorAndComponentInfo methods, minor compatibility updates

ManuelSchauer committed:
calculateDescriptor updated to handle explicit hydrogen

ManuelSchauer committed:
minor doc improvements and updated descriptors isFast flag to reflect performance benchmark results

ManuelSchauer committed:
fixed largestPiSystemDescriptor bug, minor doc improvements and updated getSpecifiedDescriptors method

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -2157,14 +1430,35 @@ void test_HYBRIDIZATION_RATIO() throws Exception {
             );
             Assertions.assertEquals(1.00, tmpMatrix[0][0]);
+        } catch (Exception anException) {
+            Assertions.fail();
```
@ManuelSchauer, personally, I usually omit any try-catch blocks in tests and just let the exception be thrown. This is equal to a test failure and displays the full exception info.

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -5513,7 +3430,7 @@ void test_Parallelization_without_Fingerprints() throws Exception {
         int tmpStartIndex = 0;
-        Descriptor[] tmpDescriptors = Descriptor.getSpecifiedDescriptors(true, true, true, false);
+        Descriptor[] tmpDescriptors = Descriptor.getAllDescriptors();
         int tmpNumberOfComponents = Descriptor.getNumberOfComponents(tmpDescriptors);
```
@ManuelSchauer, moving tests like this to a dedicated test class is a good idea we can include in the general restructuring following this PR.

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -5796,92 +3677,9 @@ void test_Integrity() throws Exception {
-            for (int i = 0; i < tmpNumberOfMolecules; i++) {
-                tmpMatrix2[i] = new float[tmpNumberOfComponents];
-            }
             aNanPositions = Collections.synchronizedList(new LinkedList<>());
```
@ManuelSchauer this is a valid point.

JonasSchaub commented:
@ManuelSchauer please update to CDK version 12 (not the snapshot) or version 13-SNAPSHOT.

JonasSchaub committed:
first part of Descriptor review, WIP

JonasSchaub committed:
WIP, does not compile, I have reached the private static methods now...

JonasSchaub committed:
continued with the review, WIP

JonasSchaub committed:
continuation of review and revision

JonasSchaub committed:
completes review of Descriptor class

JonasSchaub committed:
updates to CDK 2.13-SNAPSHOT

JonasSchaub committed:
revises exception handling in test class

JonasSchaub committed:
started individual review of test methods, WIP

JonasSchaub committed:
completes review of individual descriptor tests

JonasSchaub committed:
final part of review and revision

JonasSchaub committed:
applies spotless, so that the build succeeds

ManuelSchauer committed:
Merge pull request #10 from JonasSchaub/jonas-review

Review of entire code base

ManuelSchauer committed:
removed prefixes (a/an/tmp)

ManuelSchauer committed:
first changes of the code review:
migrate descriptor calculations to boolean error handling and minor changes

ManuelSchauer committed:
spotless fix: removed unused import

ManuelSchauer committed:
removed editor folds and added setCircularFingerprintSize method to manage circular fingerprint size

ManuelSchauer committed:
added test for getDescriptorAndComponentIndex and getDescriptorAndComponentInfo methods

ManuelSchauer committed:
added test for setCircularFingerprintSize and setFingerprintPoolSize methods

ManuelSchauer committed:
moved setDescriptorNew methods to test class and updated necessary validation helpers to package-private.
isSuccessful flag removal in setDescriptors... methods
streamlined exception variable naming
deleted 2 validation methods

ManuelSchauer committed:
removed tmp-prefixes

ManuelSchauer committed:
removed fresh molecule parsing for every ElectronDonation model
